src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,156 @@
+    private Node<T> first;
+    private Node<T> last;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> nodeToAdd = new Node<>(null, value, null);
+        if (size == 0) {
+            first = nodeToAdd;
+        } else {
+            nodeToAdd.prev = last;
+            nodeToAdd.prev.next = nodeToAdd;
+        }
+        last = nodeToAdd;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        rangeCheckForAdd(index);
+        Node<T> nodeToAdd = new Node<>(null, value, null);
+        Node<T> nodeToShift = nodeOf(index);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            first.prev = nodeToAdd;
+            nodeToAdd.next = first;
+            first = nodeToAdd;
+        }
+        nodeToShift.prev.next = nodeToAdd;
+        nodeToAdd.prev = nodeToShift.prev;
+        nodeToAdd.next = nodeToShift;
+        nodeToShift.prev = nodeToAdd;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return nodeOf(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = nodeOf(index).item;
+        nodeOf(index).item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(nodeOf(index));
     }
 
     @Override
     public boolean remove(T object) {
+        int index = 0;
+        Node<T> currentNode = first;
+        while (currentNode != null) {
+            if (currentNode.item == object || object != null && object.equals(currentNode.item)) {
+                unlink(nodeOf(index));
+                return true;
+            }
+            currentNode = currentNode.next;
+            index++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(OUT_OF_BOUNDS_EXCEPTION);
+        }
+    }
+
+    private void rangeCheckForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(OUT_OF_BOUNDS_EXCEPTION);
+        }
+    }
+
+    private Node<T> nodeOf(int index) {
```

Comment:
Let's change `nodeOf(int index)` to  `getNode(int index)`

Reply:
Ok, I'll do it.